hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -796,6 +858,476 @@ def as_result(self) -> "ConjectureResult":
+                    # Originally everything in the region 0 <= i < falsey was false
+                    # and everything above was true. We swapped one truthy element
+                    # into this region, so the region becomes 0 <= i <= falsey
+                    # except for i = 1. We know i > 1 here, so the test for truth
+                    # becomes i > falsey.
+                    result = i > falsey
+
+                if i > 1:
+                    self._cd.draw_bits(bits, forced=int(result))
+            break
+        self._cd.stop_example()
+        return result
+
+    def draw_integer(
+        self,
+        min_value: Optional[int] = None,
+        max_value: Optional[int] = None,
+        *,
+        # weights are for choosing an element index from a bounded range
+        weights: Optional[Sequence[float]] = None,
+        shrink_towards: int = 0,
+        forced: Optional[int] = None,
+    ) -> int:
+        # This is easy to build on top of our existing conjecture utils,
+        # and it's easy to build sampled_from and weighted_coin on this.
+        if weights is not None:
+            assert min_value is not None
+            assert max_value is not None
+
+            sampler = Sampler(weights)
+            idx = sampler.sample(self._cd)
+            assert shrink_towards <= min_value  # FIXME: reorder for good shrinking
+            return range(min_value, max_value + 1)[idx]
+
+        if min_value is None and max_value is None:
+            return self._draw_unbounded_integer()
+
+        if min_value is None:
+            assert max_value is not None  # make mypy happy
+            if max_value <= shrink_towards:
+                return max_value - abs(self._draw_unbounded_integer())
+            else:
+                probe = max_value + 1
+                while max_value < probe:
+                    self._cd.start_example(ONE_BOUND_INTEGERS_LABEL)
+                    probe = self._draw_unbounded_integer()
+                    self._cd.stop_example(discard=max_value < probe)
+                return probe
+
+        if max_value is None:
+            assert min_value is not None
+            if min_value >= shrink_towards:
+                return min_value + abs(self._draw_unbounded_integer())
+            else:
+                probe = min_value - 1
+                while probe < min_value:
+                    self._cd.start_example(ONE_BOUND_INTEGERS_LABEL)
+                    probe = self._draw_unbounded_integer()
+                    self._cd.stop_example(discard=probe < min_value)
+                return probe
+
+        return self._draw_bounded_integer(
+            min_value,
+            max_value,
+            center=shrink_towards,
+            forced=forced,
+        )
+
+    def draw_float(
+        self,
+        *,
+        min_value: Optional[float] = None,
+        max_value: Optional[float] = None,
+        allow_nan: bool = True,
+        smallest_nonzero_magnitude: float,
```

Comment:
opinionated change: I made `smallest_nonzero_magnitude` non-optional, as we don't know at this point whether subnormals are allowed. or not, so we can't reify a `smallest_nonzero_magnitude=None`. Happy to be overridden with an alternative.